### PR TITLE
fix(plugin): fix filling type meta and namespace in plugin reference check

### LINKED ISSUE #6290: `isRemotePluginReferenceAllowed` judges reference permission of plugin and referrer incorrectly
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

In the original specification, if a service A in namespace `ns` refers to plugin B in `ns2`, there should be a `ReferenceGrant` in `ns2` to allow reference from namespace `ns1` and group/kind of A to group/kind (and maybe name) of B: https://gateway-api.sigs.k8s.io/api-types/referencegrant/. 
Actually, if we create a `ReferenceGrant` in namespace `ns1` and allow reference from `ns1` and group/kind of A to group/kind of B, the reference check in `isRemotePluginReferenceAllowed` passes. While we create it in `ns2`, the check fails.

### Expected Behavior

The reference permission check should pass when the `ReferenceGrant` is created in the namespace of the plugin. 

### Steps To Reproduce

```markdown
1. Create a service or an ingress in namespace foo:
apiVersion: v1
kind: Service
metadata:
  namespace: foo
  name: service-foo
  annotations:
    konghq.com/plugins: bar:plugin
...
2. Create a plugin in namespace bar:
apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  namespace: bar
  name: plugin-bar
...
3. Create a ReferenceGrant in namespace bar (plugin's namespace):
apiVersion: gateway.networking.k8s.io
kind: ReferenceGrant
metadata:
  name: bar
  namespace: bar
spec:
  from:
  - group: ""
    kind: Service
    namespace: foo
  to:
  - group: "configuration.konghq.com"
    kind: KongPlugin

The plugin cannot be attached to the service. If we create the ReferenceGrant in namespace foo, the plugin can be attached.
```

### Kong Ingress Controller version

```shell
3.2.2
```

### Kubernetes version

```shell
Not releavant
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by randmonkey**
**What this PR does / why we need it**:

Correct the settings of the reference checker when we check the permission of referencing to a plugin.
The reference checker checks for if the "target" (former arg, on the "from" part of `ReferenceGrant.Spec`) is allowed to refer to "request" (latter arg, on the "to" part), so we should put the referrer on the first arg and namespace/name of plugin on the second arg for creating a reference checker.

**Which issue this PR fixes**:

fixes #6290 

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: internal/dataplane/kongstate/kongstate.go [hunk 1]
```diff
@@ -398,8 +397,14 @@ func (ks *KongState) getPluginRelations(cacheStore store.Storer, log logr.Logger
 			ingress := ks.Services[i].Routes[j].Ingress
 			pluginList := annotations.ExtractNamespacedKongPluginsFromAnnotations(ingress.Annotations)
 			for _, plugin := range pluginList {
-				// pretend we have a full Ingress struct for reference checks
+				// pretend we have a full Ingress struct for reference checks.
+				// REVIEW: we only need an object to carry type meta and object meta here, maybe we should create some other types of virtual object here?
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply internal/dataplane/kongstate/kongstate.go:401]: We can also add plugins to Kong routes translated from ``HTTPRoute` (or other `*Route` in gateway APIs) by annotation `konghq.com/plugins`. So it is possible that the `Ingress` here attached to the Kong route is not from an `Ingress`.

File: internal/dataplane/kongstate/kongstate.go [hunk 2]
```diff
@@ -398,8 +397,14 @@ func (ks *KongState) getPluginRelations(cacheStore store.Storer, log logr.Logger
...
 				virtualIngress := netv1.Ingress{
+					// Fill the actual type of the object for reference checks.
+					TypeMeta: metav1.TypeMeta{
+						APIVersion: ingress.GroupVersionKind.Group + "/" + ingress.GroupVersionKind.Version,
+						Kind:       ingress.GroupVersionKind.Kind,
+					},
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply internal/dataplane/kongstate/kongstate.go]: Added util function `TypeMetaFromGVK`.

File: internal/dataplane/kongstate/kongstate.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pmalek) [reply internal/dataplane/kongstate/kongstate.go:401]: Got it. This makes this code a bit misleading but probably due to the fact how `KongState` is structured (line 397 indicates using an `Ingress` field).

Line 398 also implies that we're using `ingress.Annotations`. If those plugin references indeed can come from other types and be saved in ingress related field(s) then this should be changed. If not in this PR then this should be logged as a separate issue.

**Comment by rainest** [comment]
Go figure, you found the same thing I did in https://github.com/Kong/kubernetes-ingress-controller/commit/dc5bd62db346c52717f94ac3ff9c16853da6788a several days earlier without my seeing it :laughing: 

Your version frames it as less of hack, which I guess alleviates some of my indecision about whether to proceed with it, though I do still think we need to rethink how the general-purpose reference checker functions work and make their intended use much clearer in the comments.

I think my confusion (and the resulting bug) came out of the original design working with reference sections in a struct (the xRoutes `backendRef` field and whatever it is that uses Secret references) that describe a target, whereas for the plugin case doesn't have an actual reference field, and I kinda wrote it thinking more about the requesting object? Not 100% sure and don't have the greatest way to describe my thinking anyway.

**Review by rainest — APPROVED** [review]
Rebased to deal with conflicts. There was nothing of note in them, just git diff logic being a bit confused by import changes and a new function in the same location, so cleared up imports and kept both test functions.

**Pull request merged**